kafka-rest/src/main/java/io/confluent/kafkarest/GroupMetadataObserver.java

```diff
@@ -0,0 +1,275 @@
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Option;
+import scala.collection.JavaConverters;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.Properties;
+import java.util.HashSet;
+import java.util.Comparator;
+
+
+public class GroupMetadataObserver {
+
+  private static KafkaConsumer<Object, Object> createConsumer(String groupId,
+                                                              KafkaRestConfig appConfig) {
+    final Properties properties = new Properties();
+    String deserializer = StringDeserializer.class.getName();
+    properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+            RestConfigUtils.bootstrapBrokers(appConfig));
+    properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+    properties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+    properties.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "30000");
+    properties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, deserializer);
+    properties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, deserializer);
+    properties.putAll(appConfig.getConsumerProperties());
+
+    return new KafkaConsumer<>(properties);
+  }
+
+  private final Logger log = LoggerFactory.getLogger(GroupMetadataObserver.class);
+
+  private final KafkaRestConfig config;
+  private final AdminClientWrapper adminClientWrapper;
+
+  public GroupMetadataObserver(KafkaRestConfig config, AdminClientWrapper adminClientWrapper) {
+    this.config = config;
+    this.adminClientWrapper = adminClientWrapper;
+  }
+
+  /**
+   * <p>Get consumer group list</p>
+   *
+   * @return list of consumer groups
+   */
+  public List<ConsumerGroup> getConsumerGroupList(Option<Integer> offsetOpt,
+                                                  Option<Integer> countOpt) throws Exception {
+    if (Option.apply(adminClientWrapper).nonEmpty()) {
+      final boolean needPartOfData = offsetOpt.nonEmpty()
+              && countOpt.nonEmpty()
+              && countOpt.get() > 0;
+      Collection<ConsumerGroupListing> groupsOverview =
+              adminClientWrapper.listConsumerGroups();
+      if (needPartOfData) {
+        final Comparator<ConsumerGroupListing> consumerGroupListingComparator =
+                Comparator.comparing(ConsumerGroupListing::groupId);
+        final List<ConsumerGroupListing> consumerGroupListings = new ArrayList<>(groupsOverview);
+        consumerGroupListings.sort(consumerGroupListingComparator);
+        groupsOverview = JavaConverters.asJavaCollection(
```

Comment:
Please, avoid reassigning variables. Mutability is hard to track and error prone. Alternative:

```
Collection<ConsumerGroupListing> groupsOverview;
if (needsPartOfData) {
    final Comparator<ConsumerGroupListing> consumerGroupListingComparator =
        Comparator.comparing(ConsumerGroupListing::groupId);
    final List<ConsumerGroupListing> consumerGroupListings = adminClientWrapper.listConsumerGroups();
    consumerGroupListings.sort(consumerGroupListingComparator);
    groupsOverview =
        JavaConverters.asJavaCollection(
            JavaConverters.collectionAsScalaIterable(consumerGroupListings)
                .slice(offsetOpt.get(), offsetOpt.get() + countOpt.get()));
} else {
    groupsOverview = adminClientWrapper.listConsumerGroups();
}
```

Now `groupsOverview` is immutable. You can even Extract-Method on the conditional block above (which I recommend).